<h2>Why is this an issue?</h2>
<p>The Dockerfile instructions are not case-sensitive. However, they should be in upper case to distinguish them from arguments more easily. Shared
coding conventions allow teams to collaborate effectively.</p>
<h3>Noncompliant code example</h3>
<pre>
from ubuntu:22.04 as jammy
</pre>
<h3>Compliant solution</h3>
<pre>
FROM ubuntu:22.04 AS jammy
</pre>

