<p>The Dockerfile instructions are not case-sensitive. However, they should be in upper case to distinguish them from arguments more easily. Shared
coding conventions allow teams to collaborate effectively.</p>
<h2>Noncompliant Code Example</h2>
<pre>
from ubuntu:22.04 as jammy
</pre>
<h2>Compliant Solution</h2>
<pre>
FROM ubuntu:22.04 AS jammy
</pre>

