<p>The variable is not available in the current scope. It will be evaluated to an empty value.</p>
<h2>Why is this an issue?</h2>
<p>The variables defined by <code>ARG</code> instruction have a scope from the definition to the end of the build stage where it was defined. If it
was defined in the beginning of the Dockerfile (outside of any build stage), then its scope is restricted to only <code>FROM</code> instructions.
Outside of their scope, variables will be resolved to empty string which may lead to unintended behaviour.</p>
<h2>How to fix it</h2>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
ARG SETTINGS
FROM busybox
RUN ./run/setup $SETTINGS
</pre>
<p>In this case the <code>$SETTINGS</code> variable will be evaluated to empty string.</p>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
FROM busybox
ARG SETTINGS
RUN ./run/setup $SETTINGS
</pre>
<p>In this case when Dockerfile will be built with the flag <code>--build-arg SETTINGS=--some-settings</code> the flag <code>--some-settings</code>
will be passed to the <code>RUN</code> instruction.</p>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
ARG SETTINGS="--default-settings"
FROM busybox
RUN ./run/setup $SETTINGS
</pre>
<p>In this case the <code>$SETTINGS</code> variable will be evaluated to empty string.</p>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
ARG SETTINGS="--default-settings"
FROM busybox
ARG SETTINGS
RUN ./run/setup $SETTINGS
</pre>
<p>In this case the flag <code>--default-settings</code> will be passed to <code>RUN</code> instruction (unless another value is provided during build
time).</p>
<h3>How does this work?</h3>
<p>The <code>FROM</code> instruction starts a new build stage where variables defined by previous <code>ARG</code> instructions are out of this new
scope. To make it accessible for the build stage they need to be defined after the <code>FROM</code> instruction.</p>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://docs.docker.com/engine/reference/builder/#scope">ARG scope - Dockerfile reference</a> </li>
</ul>

