<p>Various package and software management applications require manual input for execution confirmation by default. This confirmation is usually
required when installing, updating, or removing programs and packages. To execute the command without manual input, a general consent can be
given.</p>
<h2>Why is this an issue?</h2>
<p>If such a command is part of a script that is executed automatically, the execution is aborted because there is no confirming manual input.
Instructions, such as installing programs or updating them, cannot be performed in an automated way. This applies, among others, to the advanced
package tool, 'apt-get'.</p>
<h2>How to fix it</h2>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
RUN apk add ca-certificates
RUN apt-get install ca-certificates
RUN aptitude install ca-certificates
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
RUN apk -y add ca-certificates
RUN apt-get install -y ca-certificates
RUN aptitude -y install ca-certificates
</pre>
<h3>How does this work?</h3>
<p>If the <code>-y</code> flag set, no manual input is expected, and it is assumed that there is general consent. For <code>apk</code> and
<code>apt-get</code> also the long versions <code>--yes</code> and <code>--assume-yes</code> exist. For <code>aptitude</code> also the long version
<code>--assume-yes</code> exist.</p>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://linux.die.net/man/8/apt-get">apt-get - Linux man page</a> </li>
  <li> <a href="https://docs.alpinelinux.org/user-handbook/0.1a/Working/apk.html">apk - Alpine Package Keeper</a> </li>
  <li> <a href="https://wiki.debian.org/Aptitude">aptitude - Linux aptitude command</a> ] </li>
  <li> <a href="https://docs.docker.com/engine/reference/builder/#run">RUN - Docker reference</a> </li>
  <li> <a href="https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run">RUN - Best practices for writing Dockerfiles</a> </li>
</ul>

