<p>Deprecated instructions should be replaced by other suggested instructions.</p>
<h2>Why is this an issue?</h2>
<p>Languages and technologies evolve, which leads to deprecation of some features. At some day they may be removed. It is important to use up-to-date
constructs to avoid issues by migration to newer version of technology.</p>
<h2>How to fix it</h2>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
MAINTAINER bob
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
LABEL org.opencontainers.image.authors="bob"
</pre>
<h3>How does this work?</h3>
<p>The <code>LABEL</code> instruction is much more flexible than <code>MAINTAINER</code> and should be used instead.</p>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://docs.docker.com/engine/deprecated/#maintainer-in-dockerfile">Deprecated Engine Features</a> </li>
  <li> <a href="https://docs.docker.com/engine/reference/builder/#maintainer-deprecated">MAINTAINER (deprecated)ðŸ”— - Dockerfile reference</a> </li>
</ul>

