<p>The Dockerfile should contain at most one <code>ENTRYPOINT</code> and one <code>CMD</code> instruction, because only the last one will have an
effect.</p>
<h2>Why is this an issue?</h2>
<p>Multiple <code>ENTRYPOINT</code> or <code>CMD</code> instructions in a file can lead to confusion as we may think they are all applied. This is not
the case, as only the last one is applied.</p>
<h2>How to fix it</h2>
<p>Previous <code>ENTRYPOINT</code> and <code>CMD</code> instructions should be removed to avoid this.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
FROM busybox
ENTRYPOINT ignored_entrypoint param1 param2
ENTRYPOINT effective_entrypoint param1 param2

CMD ignored_cmd param1 param2
CMD effective_cmd param1 param2
</pre>
<p>Here we have multiple <code>ENTRYPOINT</code> and <code>CMD</code> instructions. The first <code>ENTRYPOINT</code> and the first <code>CMD</code>
instructions will have no effect. Although this is valid in Docker, this can lead to confusion and be error-prone, as we may expect each
<code>CMD</code> and <code>ENTRYPOINT</code> to have an effect.</p>
<p>Multi-Stage Build:</p>
<pre data-diff-id="2" data-diff-type="noncompliant">
FROM scratch as development
CMD ignored_scratch_cmd param1 param2
CMD effective_scratch_cmd param1 param2

FROM busybox
CMD ignored_busyBox_cmd param1 param2
CMD effective_busyBox_cmd param1 param2
</pre>
<p>For multi-stage builds we take each stage into account separately. This is because there are valid docker setups, where the file should only be
build up to a certain stage. In the example, the developer builds only the first stage as a development environment via <code>docker build --target
development</code>.</p>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
FROM busybox
ENTRYPOINT effective_entrypoint param1 param2

CMD effective_cmd param1 param2
</pre>
<p>Here we have only one <code>ENTRYPOINT</code> and one <code>CMD</code> instruction. Each of them will be considered by the docker container and
have a normal effect, as we can expect.</p>
<p>Multi-Stage Build:</p>
<pre data-diff-id="2" data-diff-type="compliant">
FROM scratch as development
CMD effective_scratch_cmd param1 param2

FROM busybox
CMD effective_busyBox_cmd param1 param2
</pre>
<p>For each stage, we only have one <code>CMD</code> or <code>ENTRYPOINT</code> instruction.</p>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://docs.docker.com/engine/reference/builder/#entrypoint">ENTRYPOINT - Dockerfile reference</a> </li>
  <li> <a href="https://docs.docker.com/engine/reference/builder/#cmd">CMD - Dockerfile reference</a> </li>
  <li> <a href="https://docs.docker.com/build/building/multi-stage/#stop-at-a-specific-build-stage">Multi-Stage builds - Stop at specific build
  stage</a> </li>
</ul>

