<h2>Why is this an issue?</h2>
<p>In Dockerfile, instructions <code>RUN</code>, <code>CMD</code>, and <code>ENTRYPOINT</code> can contain long shell scripts chaining multiple
commands, including the <code>cd</code> command for changing directories. Using <code>WORKDIR</code> instruction instead reduces the complexity of the
above instructions and makes them easier to read, understand, troubleshoot, and maintain.</p>
<h3>What is the potential impact?</h3>
<p>The Dockerfile instructions like <code>RUN</code> allow users to execute longer scripts. See the following example:</p>
<pre>
RUN cd /tmp &amp;&amp; \
  git clone myrepository.com/MyOrganization/project &amp;&amp; \
  cd project &amp;&amp; \
  make &amp;&amp; \
  cd /app/bin
</pre>
<p>In this example, the first <code>cd /tmp</code> command can be replaced by <code>WORKDIR /tmp</code> before <code>RUN</code>. The last <code>cd
/app/bin</code> command can be replaced with <code>WORKDIR /app/bin</code> after the <code>RUN</code> instruction. The result will be the
following:</p>
<pre>
WORKDIR /tmp
RUN git clone myrepository.com/MyOrganization/project &amp;&amp; \
  cd project &amp;&amp; \
  make
WORKDIR /app/bin
</pre>
<p>Those actions will reduce the length of the <code>RUN</code> instruction, which makes it easier to read and understand. Sometimes, it is hard to
avoid the usage of <code>cd</code> command, especially in the middle of a long script. Removing them from the beginning and end of a multi-line is an
easy improvement. Additionally, many commands work well with absolute paths, so changing directories can be avoided in most cases.</p>
<p>The <code>WORKDIR</code> instruction can be used multiple times in a Dockerfile. It changes the current directory for the next instructions and
until there is a following change. This approach simplifies understanding of what is a current directory.</p>
<p>The same principles apply to <code>CMD</code> and <code>ENTRYPOINT</code> instructions.</p>
<p>This recommendation provides a clear structure for Dockerfiles, making it easier to maintain.</p>
<h2>How to fix it</h2>
<p>Where possible, ensure that all usages of <code>cd</code> commands are replaced by a <code>WORKDIR</code> instruction. The <code>cd</code> commands
at the beginning or end of a chain of commands are a reliable sign that they can be replaced. Also, using absolute paths can be considered for
commands that accept them.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
RUN cd /app/bin &amp;&amp; ./start.sh
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
WORKDIR /app/bin
RUN ./start.sh

# Or:
RUN /app/bin/start.sh
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> <a href="https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#workdir">WORKDIR - Best practices for writing
  Dockerfiles</a> </li>
  <li> <a href="https://docs.docker.com/engine/reference/builder/#workdir">WORKDIR - Dockerfile reference</a> </li>
</ul>

