<p>This rule raise an issue when one of the mandatory label are missing.</p>
<h2>Why is this an issue?</h2>
<p>Adding labels to your image help to organize images by project, record licensing information, aid in automation and for other reasons.<br> The rule
provide the possibility to configure the list of mandatory label that must be present in every Dockerfile.</p>
<h2>How to fix it</h2>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<p>With <code>labels</code> set to <code>maintainer,description,version</code></p>
<pre data-diff-id="1" data-diff-type="noncompliant">
FROM ubuntu:22.02
RUN my_command
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
FROM ubuntu:22.02
LABEL maintainer="John Doe"
LABEL description="Image doing XYZ"
LABEL version=1.0
RUN my_command
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Docker docs - <a href="https://docs.docker.com/build/building/best-practices/#label">Building best practices - LABEL</a> </li>
</ul>

