<p>In Dockerfiles, commands can be specified either as a single string or as a JSON array of strings. The latter is called an "exec form". However,
when exec form is not a valid JSON, it will be silently treated as shell form. Usually, this will lead to a crash, but sometimes it can pass silently
and lead to unexpected behavior.</p>
<h2>Why is this an issue?</h2>
<p>Some tricky issues with exec form include having symbols after the closing bracket (silently passes with Docker earlier than 27.1.0) or using
incorrect symbol for quotes within a JSON array. In these cases, the exec form will be treated as shell form, and the build will pass, but the command
will not be executed as expected, because the whole string will be passed to a shell.</p>
<h2>How to fix it</h2>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
FROM alpine:3.14
ENTRYPOINT ["/app", "-c", "config.json"] -D
</pre>
<pre data-diff-id="2" data-diff-type="noncompliant">
FROM alpine:3.14
ENTRYPOINT [ "/bin/bash”, “-c” ]
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
FROM alpine:3.14
ENTRYPOINT ["/app", "-c", "config.json", "-D"]
</pre>
<pre data-diff-id="2" data-diff-type="compliant">
FROM alpine:3.14
ENTRYPOINT [ "/bin/bash", "-c" ]
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Docker Docs - <a href="https://docs.docker.com/reference/dockerfile/#shell-and-exec-form">Shell and exec form</a> </li>
</ul>

