<p>This rule raises an issue when template directives do not have a whitespace after the opening brace and before the closing brace. Itâ€™s based on
best practices in Helm and emphasizes the importance of maintaining this whitespace for better readability and understanding of the code.</p>
<h2>Why is this an issue?</h2>
<p>The absence of whitespace in template directives in Helm can lead to several issues:</p>
<p><strong>Readability</strong>: Code is read more often than it is written. When template directives are tightly packed without whitespaces, it can
be challenging to distinguish between the braces and the content inside. This can slow down the process of understanding the code, especially for
those who are new to the codebase.</p>
<p><strong>Maintainability</strong>: Hard-to-read code is also hard to maintain. If a developer struggles to understand what a piece of code does due
to poor formatting, they may introduce bugs when they try to modify it.</p>
<p>In summary, not having a whitespace after the opening brace and before the closing brace in template directives can make the code harder to read,
understand, and maintain.</p>
<h2>How to fix it in Helm</h2>
<p>To fix this issue, add a whitespace after the opening brace and before the closing brace in template directives. Also, ensure there is a whitespace
after opening chomping braces <code>{{-</code> and before closing chomping braces <code>-}}</code>.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
apiVersion: v1
kind: Pod
metadata:
  name: {{.Values.podName}} # Noncompliant
spec:
  containers:
  - name: {{-.Values.containerName-}} # Noncompliant
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
apiVersion: v1
kind: Pod
metadata:
  name: {{ .Values.podName }}
spec:
  containers:
  - name: {{- .Values.containerName -}}
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Helm Best Practices - <a href="https://helm.sh/docs/chart_best_practices/templates/#formatting-templates">Formatting Templates</a> </li>
  <li> Go Documentation - <a href="https://pkg.go.dev/text/template?utm_source=godoc#hdr-Text_and_spaces">Text and spaces</a> </li>
</ul>

