<p>When deploying an Azure Resource Manager template (ARM template), you must provide a location for each resource. Locations can either be added in a
template directly or passed as a parameter. Itâ€™s advisable to use a parameter to specify the location of resources.</p>
<h2>Why is this an issue?</h2>
<p>When deploying an Azure Resource Manager template (ARM template), you must provide a location for each resource. This can be done directly in the
template or by passing parameters. However, hardcoding locations in the template can limit flexibility and potentially create deployment challenges,
restricting users from choosing their preferred deployment location.</p>
<p>It is therefore recommended to use a parameter to specify the location for resources, with the default value set to
<code>resourceGroup().location</code>. This practice ensures consistency in resource allocation and provides users of the template the flexibility to
specify a location where they have the necessary permissions to deploy resources. This approach helps avoid hardcoding locations, which can lead to
potential deployment issues and restrictions.</p>
<h2>How to fix it in ARM templates</h2>
<p>Create a parameter for the location and set the default value to <code>resourceGroup().location</code>. Then, use the parameter to specify the
location of resources.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="11" data-diff-type="noncompliant">
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "apiVersion": "2022-09-01",
      "name": "[parameters('storageAccountName')]",
      "location": "westus",
      "sku": {
        "name": "Standard_LRS"
      },
      "kind": "StorageV2"
    }
  ]
}
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="11" data-diff-type="compliant">
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "location": {
      "type": "string",
      "defaultValue": "[resourceGroup().location]",
      "metadata": {
        "description": "Location for all resources."
      }
    }
  },
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "apiVersion": "2022-09-01",
      "name": "[parameters('storageAccountName')]",
      "location": "[parameters('location')]",
      "sku": {
        "name": "Standard_LRS"
      },
      "kind": "StorageV2"
    }
  ]
}
</pre>
<h2>How to fix it in Bicep</h2>
<p>Create a parameter for the location and set the default value to <code>resourceGroup().location</code>. Then, use the parameter to specify the
location of resources.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
resource storageAccount 'Microsoft.Storage/storageAccounts@2022-09-01' = {
  name: storageAccountName
  location: 'westus'
  sku: {
    name: 'Standard_LRS'
  }
  kind: 'StorageV2'
}
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
param location string = resourceGroup().location

resource storageAccount 'Microsoft.Storage/storageAccounts@2022-09-01' = {
  name: storageAccountName
  location: location
  sku: {
    name: 'Standard_LRS'
  }
  kind: 'StorageV2'
}
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Azure Resource Manager - <a href="https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/resource-location">Resource location
  in templates</a> </li>
</ul>

