<h2>Why is this an issue?</h2>
<p>An unused parameter is a parameter that has been declared but is not used anywhere in the block of code where it is defined. It is dead code,
contributing to unnecessary complexity and leading to confusion when reading the code. Therefore, it should be removed from your code to maintain
clarity and efficiency.</p>
<h3>What is the potential impact?</h3>
<p>Having unused parameters in your code can lead to several issues:</p>
<ul>
  <li> <strong>Decreased Readability</strong>: Unused parameters can make your code more difficult to read. They add extra lines and complexity, which
  can distract from the main logic of the code. </li>
  <li> <strong>Misunderstanding</strong>: When other developers read your code, they may wonder why a variable is declared but not used. This can lead
  to confusion and misinterpretation of the codeâ€™s intent. </li>
  <li> <strong>Potential for Bugs</strong>: If a parameter is declared but not used, it might indicate a bug or incomplete code. For example, if you
  declared a parameter intending to use it in a calculation, but then forgot to do so, your program might not work as expected. </li>
  <li> <strong>Maintenance Issues</strong>: Unused parameters can make code maintenance more difficult. If a programmer sees an unused parameter, they
  might think it is a mistake and try to 'fix' the code, potentially introducing new bugs. </li>
</ul>
<p>In summary, unused local parameters can make your code less readable, more confusing, and harder to maintain, and they can potentially lead to
bugs. Therefore, it is best to remove them.</p>
<h2>How to fix it in ARM Templates</h2>
<p>The fix for this issue is straightforward. Once you ensure the unused parameter is not part of an incomplete implementation leading to bugs, you
just need to remove it.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="1" data-diff-type="noncompliant">
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "parameters": {
      "unusedParameter": {
          "type": "string"
      },
      "virtualMachinesName": {
          "type": "string"
      }
  },
  "resources": [
    {
      "type": "Microsoft.Compute/virtualMachines",
      "name": "[parameters('virtualMachinesName')]",
      "apiVersion": "2023-09-01",
      "location": "[resourceGroup().location]"
    }
  ]
}
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="1" data-diff-type="compliant">
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "parameters": {
      "virtualMachinesName": {
          "type": "string"
      }
  },
  "resources": [
    {
      "type": "Microsoft.Compute/virtualMachines",
      "name": "[parameters('virtualMachinesName')]",
      "apiVersion": "2023-09-01",
      "location": "[resourceGroup().location]"
    }
  ]
}
</pre>
<h2>How to fix it in Bicep</h2>
<p>The fix for this issue is straightforward. Once you ensure the unused parameter is not part of an incomplete implementation leading to bugs, you
just need to remove it.</p>
<h3>Code examples</h3>
<h4>Noncompliant code example</h4>
<pre data-diff-id="2" data-diff-type="noncompliant">
param unusedParameter string // Noncompliant
param virtualMachinesName string = '${uniqueString(resourceGroup().id)}'

resource vm 'Microsoft.Compute/virtualMachines@2023-09-01' = {
  name: virtualMachinesName
  location: resourceGroup().location
}
</pre>
<h4>Compliant solution</h4>
<pre data-diff-id="2" data-diff-type="compliant">
param virtualMachinesName string = '${uniqueString(resourceGroup().id)}'

resource vm 'Microsoft.Compute/virtualMachines@2023-09-01' = {
  name: virtualMachinesName
  location: resourceGroup().location
}
</pre>
<h2>Resources</h2>
<h3>Documentation</h3>
<ul>
  <li> Microsoft - <a href="https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/best-practices#parameters">ARM template best
  practices</a> </li>
  <li> Microsoft - <a href="https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/best-practices#names">Best practices for Bicep</a>
  </li>
  <li> Microsoft - <a href="https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/parameters">Parameters in ARM templates</a> </li>
  <li> Microsoft - <a href="https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/parameters">Parameters in Bicep</a> </li>
</ul>

