<h2>Why is this an issue?</h2>
<p>Shared conventions allow teams to collaborate effectively. This rule allows to check that all tag keys match a provided regular expression.</p>
<h3>Noncompliant code example</h3>
<p>With default provided regular expression ^([A-Z]<strong>:)</strong>([A-Z][A-Za-z]*)$:</p>
<pre>
AWSTemplateFormatVersion: 2010-09-09
Resources:
  S3Bucket:
    Type: 'AWS::S3::Bucket'
    Properties:
      BucketName: "mybucketname"
      Tags:
        - Key: "anycompany:cost-center" # Noncompliant
          Value: "Accounting"
        - Key: "anycompany:EnvironmentType" # Noncompliant
          Value: "PROD"
</pre>
<p>&nbsp;</p>
<h3>Compliant solution</h3>
<p>&nbsp;</p>
<pre>
AWSTemplateFormatVersion: 2010-09-09
Resources:
  S3Bucket:
    Type: 'AWS::S3::Bucket'
    Properties:
      BucketName: "mybucketname"
      Tags:
        - Key: "Anycompany:CostCenter"
          Value: "Accounting"
        - Key: "Anycompany:EnvironmentType"
          Value: "PROD"
</pre>
<h2>Resources</h2>
<ul>
  <li> <a href="https://docs.aws.amazon.com/whitepapers/latest/tagging-best-practices/adopt-a-standardized-approach-for-tag-names.html">AWS
  Documentation</a>: Adopt a Standardized Approach for Tag Names </li>
</ul>

